tools/gfx/d3d11/render-d3d11.cpp

```diff
@@ -542,22 +423,1101 @@ class D3D11Device : public ImmediateRendererBase
+            //
+            RefPtr<ShaderObjectLayoutImpl> specializedLayout;
+            SLANG_RETURN_ON_FAIL(_getSpecializedLayout(specializedLayout.writeRef()));
+
+            auto specializedOrdinaryDataSize = specializedLayout->getElementTypeLayout()->getSize();
+            if (specializedOrdinaryDataSize == 0)
+                return SLANG_OK;
+
+            // Once we have computed how large the buffer should be, we can allocate
+            // it using the existing public `IDevice` API.
+            //
+
+            ComPtr<IBufferResource> bufferResourcePtr;
+            IBufferResource::Desc bufferDesc;
+            bufferDesc.init(specializedOrdinaryDataSize);
+            bufferDesc.cpuAccessFlags |= IResource::AccessFlag::Write;
+            SLANG_RETURN_ON_FAIL(device->createBufferResource(
+                IResource::Usage::ConstantBuffer, bufferDesc, nullptr, bufferResourcePtr.writeRef()));
+            m_ordinaryDataBuffer = static_cast<BufferResourceImpl*>(bufferResourcePtr.get());
+
+            // Once the buffer is allocated, we can use `_writeOrdinaryData` to fill it in.
+            //
+            // Note that `_writeOrdinaryData` is potentially recursive in the case
+            // where this object contains interface/existential-type fields, so we
+            // don't need or want to inline it into this call site.
+            //
+            SLANG_RETURN_ON_FAIL(_writeOrdinaryData(device, m_ordinaryDataBuffer, 0, specializedOrdinaryDataSize, specializedLayout));
+
+            return SLANG_OK;
+        }
+
+        /// Bind the buffer for ordinary/uniform data, if needed
+        Result _bindOrdinaryDataBufferIfNeeded(
+            D3D11Device* device,
+            RootBindingState* bindingState)
+        {
+            // We start by ensuring that the buffer is created, if it is needed.
+            //
+            SLANG_RETURN_ON_FAIL(_ensureOrdinaryDataBufferCreatedIfNeeded(device));
+
+            // If we did indeed need/create a buffer, then we must bind it
+            // into root binding state.
+            //
+            if (m_ordinaryDataBuffer)
+            {
+                bindingState->constantBuffers.add(m_ordinaryDataBuffer->m_buffer);
+            }
+
+            return SLANG_OK;
+        }
+    public:
+        virtual Result bindObject(D3D11Device* device, RootBindingState* bindingState)
+        {
+            ShaderObjectLayoutImpl* layout = getLayout();
+
+            Index baseRangeIndex = 0;
+            SLANG_RETURN_ON_FAIL(_bindOrdinaryDataBufferIfNeeded(device, bindingState));
+
+            for (auto sampler : m_samplers)
+                bindingState->samplerBindings.add(sampler ? sampler->m_sampler.get() : nullptr);
+            
+            for (auto srv : m_srvs)
+                bindingState->srvBindings.add(srv ? srv->m_srv : nullptr);
+
+            for (auto uav : m_uavs)
+                bindingState->uavBindings.add(uav ? uav->m_uav : nullptr);
```

Comment:
In the hypothetical model where this operation directly talks to the `ID3D11DeviceContext`, each of these loops could be a single `CSSet*()` call in the compute case. We'd just need to store parallel arrays that are already in D3D11-API-friendly format, list a `List<ID3D11ShaderResourceView*>`.

I think having a really clear operation on a shader object that shows how it "splats" its state onto the context in as few operations as possible will go a long way to showing that they are actually a highly efficient way to drive the API.

Reply:
You are right, I didn't realize you can call *SSetXXX multiple times, since all UAVs needs to be set at a single time, but that is just an anomaly.

Reply:
Oh yeah, I forgot that UAVs were a special case in D3D11 (and an even more special case when it comes to RTVS + UAVs for PS).

One thing that I guess I forgot to note in the original review is that it is kind of subtle but this logic in the D3D11 handling of shader objects seems to be assuming that applications only use the completely default layout rules (no explicit `register` stuff). We probaby need/want to address that sooner or later if we want to have tests that also make use of explicit bindings (even if they are basically always a Bad Idea).